hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/execution/traceability/HederaTracer.java

```diff
@@ -68,11 +77,56 @@ public HederaTracer(final boolean areActionSidecarsEnabled) {
 
     @Override
     public void init(final MessageFrame initialFrame) {
+        log.atInfo().log("** at HederaTracer.init");
         if (areActionSidecarsEnabled) {
             trackTopLevelActionFor(initialFrame);
         }
     }
 
+    @Override
+    public void finalizeOperation(final MessageFrame initialFrame) {
+        if (areActionSidecarsEnabled) {
+            // make sure only valid actions are left in allActions
+            final var validActions =
+                    allActions.stream().filter(SolidityAction::isValid).toList();
+
+            if (!currentActionsStack.isEmpty() || validActions.size() != allActions.size()) {
```

Comment:
There's a state machine here that runs through these states.  Unfortunately, it is not actually "here".  It's over somewhere in the EVM machinery and all we have here are these callbacks for certain transitions - and it isn't even clear at this point what happens in what order, or what could interrupt the state machine's progress over there in the EVM machinery.  At least as far as I can see.  (Because: nothing's documented that I can point to.)

So I don't _expect_ there's anything wrong at all.  But I can't show that by pointing to a state diagram and mapping things out.  And if there's a `pop` and there's nothing to be popped - that's a hard error (throws an exception).  As long as I'm making things safe for field values (by validation and tossing records that don't work) I thought I'd check for this too.  And log in both cases.

I will add a comment as suggested once I figure out how to word it.